File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        ArrayList<String> list = new ArrayList<>();

        for (int i=0; i<want.length; i++) {
            while (number[i]-- > 0) {
                list.add(want[i]);
            }
        }

        for (int i=0; i<discount.length-list.size()+1;i ++) {
            ArrayList<String> l = (ArrayList<String>) list.clone();
            for (int j=i; j<i+list.size(); j++) {
                if (l.contains(discount[j])) {
                    l.remove(discount[j]);
                } else {
                    break;
                }
            }
            answer += l.size()==0 ? 1 : 0;
        }

        return answer;
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
function solution(want, number, discount) {
    var answer = 0;
    const memo = []
    for (let i = 0; i < want.length; i++) {
        memo.push([want[i], number[i]])
    }
    
    for (let i = 0; i + 10 <= discount.length; i++) {
        const cart = new Map(memo);
        for (let j = i; j < i + 10; j++) {
            if (cart.has(discount[j])) {
                cart.set(discount[j], cart.get(discount[j]) - 1)
            }
        }
        if ([...cart.values()].filter(m => m > 0).length === 0) {
            answer++;
        }
    }
    
    return answer;
}